CMakeLists.txt

```diff
@@ -11,6 +11,21 @@ project(
     LANGUAGES C CXX
 )
 
+MACRO(HEADER_DIRECTORIES return_list)
+    FILE(GLOB_RECURSE new_list *.hpp)
+    SET(dir_list "")
+    FOREACH(file_path ${new_list})
+        GET_FILENAME_COMPONENT(dir_path ${file_path} PATH)
+        SET(dir_list ${dir_list} ${dir_path})
+    ENDFOREACH()
+    LIST(REMOVE_DUPLICATES dir_list)
+    SET(${return_list} ${dir_list})
+ENDMACRO()
+
+HEADER_DIRECTORIES(header_list)
+
+include_directories(${header_list})
+
```

Comment:
Adding all directories as include directories seems to go against the purpose of having these directories.

I imagined this would follow the [Google C++ Guidelines for includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) so only the `src` dir would need to be included, and that the directory names would be removed from the moved files.

Reply:
I was going to remove the directory names from the files once they were refactored into new namespaces on a case by case basis so we can track what has and has not been refactored yet

Reply:
That works for me.

Reply:
In terms of adding all the dirs. I was trying to avoid includes needing to do the following #include "../files/file.hpp" and instead can just do #include "file.hpp" the idea being that if a file is moved to a new folder dir in the future each #include where that file is used doesn't need to be updated

Reply:
I can update this to conform to the google C++ guideline though

Reply:
I want to know the directory of the file being included.  I like how Google does it, they ban `../` from project includes so it won't be hard to change their locations in the future.  This can be done later.

Reply:
> I can update this to conform to the google C++ guideline though

This project doesn't have set guidelines yet, but I like to have some and I tend to fallback on Google's guidelines when some kind of issue comes up that a guideline would solve.